File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/graphql/ProductConnection.java
package com.zappymart.adapters.graphql;

import com.zappymart.application.Page;
import com.zappymart.domain.catalogue.Product;

import java.util.List;

public record ProductConnection(List<ProductEdge> edges, PageInfo pageInfo, int totalCount) {

    public static ProductConnection of(Page<Product> page) {
        List<ProductEdge> edges = page.items().stream()
                .map(product -> new ProductEdge(Cursors.encode(product.id()), product))
                .toList();
        String endCursor = edges.isEmpty() ? null : edges.getLast().cursor();
        return new ProductConnection(edges, new PageInfo(page.hasNextPage(), endCursor), page.totalCount());
    }
}

File: backends/dotnet/src/Zappy.Adapters.GraphQL/Types/ProductConnection.cs
using Zappy.Application;
using Zappy.Domain;

namespace Zappy.Adapters.GraphQL;

public sealed class ProductConnection(IReadOnlyList<ProductEdge> edges, PageInfo pageInfo, int totalCount)
{
    public IReadOnlyList<ProductEdge> Edges { get; } = edges;

    public PageInfo PageInfo { get; } = pageInfo;

    public int TotalCount { get; } = totalCount;

    public static ProductConnection From(Page<Product> page)
    {
        var edges = page.Items.Select(product => new ProductEdge(Cursor.For(product.Id), product)).ToList();
        return new ProductConnection(
            edges,
            new PageInfo(page.HasNextPage, edges.Count == 0 ? null : edges[^1].Cursor),
            page.TotalCount);
    }
}
